### ISSUE TITLE
run-example issue

### ISSUE THREAD

**Issue by freealong** [nmellado/Super4PCS#36]
I try to run the example
```bash
./run-example.sh
```
The output:
```
Use Super4PCS
norm_max_dist: 0.010000
Work with 200 points
Initial LCP: 0.010000
Score: 0.01est: 0.010000                  
(Homogeneous) Transformation from hippo2.obj to hippo1.obj:
1 0 0 0
0 1 0 0
0 0 1 0
0 0 0 1
Exporting Matrix to mat_super4pcs_fast.txt...DONE
Exporting Registered geometry to super4pcs_fast.obj...DONE
real 0.85
user 0.78
sys 0.07
Use old 4PCS
norm_max_dist: 0.010000
Initial LCP: 0.010000
Score: 0.535st: 0.535000                  
(Homogeneous) Transformation from hippo2.obj to hippo1.obj:
  0.780739 -0.0319342   -0.62404 -0.0981111
-0.0509506   0.992114  -0.114514 -0.0217299
  0.622776   0.121201   0.772955 -0.0271737
         0          0          0          1
Exporting Matrix to mat_4pcs_fast.txt...DONE
Exporting Registered geometry to 4pcs_fast.obj...DONE
real 17.72
user 17.63
sys 0.08
```
It seem the Super4PCS's Transformation is wrong...

**Comment by nmellado** [comment]
Yes indeed...
Could you please add more details about your OS, compiler, etc ?

Thanks for the report.

**Comment by freealong** [comment]
OS：Ubuntu 16.04 x64
Compiler：clang 3.8

发自我的 iPhone

**Comment by nmellado** [comment]
I cannot reproduce this behavior with Ubuntu 17.04 running with clang 3.8.1.
Have you set any CMake option, such as `SUPER4PCS_USE_CHEALPIX` ? 

In order to understand what is going on, could you please run the following commands and past the output here:
```
make buildtests
make test
```

Thanks

**Comment by freealong** [comment]
I did not set any CMake options. Just clone the projects and
```bash
mkdir build
cd build
cmake ..
make
```

make test output as follow:
```bash
Test project /home/yongqi/projects/Super4PCS/build
    Start 1: pair_extraction
1/4 Test #1: pair_extraction ..................   Passed   14.28 sec
    Start 2: quad_extraction
2/4 Test #2: quad_extraction ..................***Exception: Other  0.86 sec
    Start 3: matching_0
3/4 Test #3: matching_0 .......................***Exception: Other  4.73 sec
    Start 4: matching_1
4/4 Test #4: matching_1 .......................***Exception: Other  3.42 sec

25% tests passed, 3 tests failed out of 4

Total Test time (real) =  23.33 sec

The following tests FAILED:
	  2 - quad_extraction (OTHER_FAULT)
	  3 - matching_0 (OTHER_FAULT)
	  4 - matching_1 (OTHER_FAULT)
Errors while running CTest
Makefile:83: recipe for target 'test' failed
make: *** [test] Error 8
```

Besides， I rebuild the project with gcc-5.4. The example's output doesn't change.

**Comment by Gregoriomartjim** [comment]
I am having the same problem here i am getting this problem
time: cannot run ./build/Super4PCS: No such file or directory
Command exited with non-zero status 127
real 0.00
user 0.00
sys 0.00
time: cannot run ./build/Super4PCS: No such file or directory
Command exited with non-zero status 127
real 0.00
user 0.00
sys 0.00

**Comment by freealong** [comment]
@Gregoriomartjim I guess there is no Super4PCS under your build directory. Maybe your build process failed.

**Comment by nmellado** [comment]
@freealong , could you please try to run Super4PCS using more samples, for instance using the following command:
```
time -p ./build/Super4PCS -i hippo1.obj hippo2.obj -o 0.7 -d 0.01 -t 1000 -n 400 -r super4pcs_fast.obj -m mat_super4pcs_fast.txt
```

@Gregoriomartjim As @freealong said, you probably have a build problem. Please open a dedicated issue.

**Comment by freealong** [comment]
I changed the samples and ran the following command:
```
./build/Super4PCS -i ArmadilloBack_0.ply ArmadilloBack_30.ply -o 0.7 -d 0.01 -t 1000 -n 400 -r super4pcs_fast.obj -m mat_super4pcs_fast.txt
```
The output is right.

Then, I use Meshlab to transform ArmadilloBack_0.ply into a new ply(ArmadilloBack_1.ply) and run Super4PCS again:
 ```
./build/Super4PCS -i ArmadilloBack_0.ply ArmadilloBack_30.ply -o 0.7 -d 0.01 -t 1000 -n 400 -r super4pcs_fast.obj -m mat_super4pcs_fast.txt
```
The result homogeneous transformation is Identity. The output is wrong.

**Comment by nmellado** [comment]
Thanks @freealong for this test.

Could you please save the subsampled clouds (using `--sampled1` and `--sampled2`) in both cases and check if they are similar ?
I suspect your bug comes from the sampling stage...

Also, can you also run and paste here the output of the following commands ?
```
cd build
make dl-datasets
cd datasets/demo1
./run.sh
```
A normal run should not produce any registration error.

**Comment by freealong** [comment]
In the first case (runs successfully) the two sampled clouds are not similar.
In the second case the two sampled clouds are quiet similar. 
It seems the bug do come from the sampling stage.

Command './run.sh' output:
```
$ ../../Super4PCS -i bird/frame_0000.ply bird/frame_0002.ply -d 3    -o 0.8  -t 10000 -n 500  -r bird/frame_0002_aligned.ply
Use Super4PCS
norm_max_dist: 3.000000
Work with 650 points
Initial LCP: 0.060000
Score: 0.533846
(Homogeneous) Transformation from bird/frame_0002.ply to bird/frame_0000.ply:
  0.22887 -0.930223 -0.286887  -738.253
 0.137925 -0.260748  0.955503    -242.4
-0.963636 -0.258255  0.068624   527.881
        0         0         0         1
Exporting Registered geometry to bird/frame_0002_aligned.ply...DONE
real 21.92
user 21.88
sys 0.03
$ ../../Super4PCS -i buba/buba1.obj  buba/buba2.obj          -d 1    -o 0.95 -t 10000 -n 1000 -r buba/buba2_aligned.obj
Use Super4PCS
norm_max_dist: 1.000000
Work with 1083 points
Initial LCP: 0.037858
Score: 0.0904894
(Homogeneous) Transformation from buba/buba2.obj to buba/buba1.obj:
  0.999393 -0.0311002  0.0157263   -1.99853
-0.0327566  -0.992323   0.119259    1.13148
 0.0118967  -0.119702  -0.992738       3.32
         0          0          0          1
Exporting Registered geometry to buba/buba2_aligned.obj...DONE
real 0.52
user 0.40
sys 0.11
$ ../../Super4PCS -i head/input1.obj head/input2.obj         -d 2    -o 0.6  -t 10000 -n 1500 -r head/input2_aligned.obj
Use Super4PCS
norm_max_dist: 2.000000
Work with 1656 points
Initial LCP: 0.051932
Score: 0.267512
(Homogeneous) Transformation from head/input2.obj to head/input1.obj:
  0.958306  -0.283543 -0.0353954    52.0947
  0.159462   0.633462  -0.757164    494.348
   0.23711   0.719951   0.652265    181.109
         0          0          0          1
Exporting Registered geometry to head/input2_aligned.obj...DONE
real 10.50
user 10.02
sys 0.47
$ ../../Super4PCS -i kinect/guy/1.ply kinect/guy/2.ply       -d 0.03 -o 0.4  -t 10000 -n 1000 -c 200 -r kinect/guy/2_aligned.ply
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1030 points
Initial LCP: 0.052427
Score: 0.463107
(Homogeneous) Transformation from kinect/guy/2.ply to kinect/guy/1.ply:
  0.860544 -0.0214595   0.508924  -0.865163
-0.0628261   0.987012   0.147852  -0.269905
 -0.505487  -0.159207   0.848019   0.313234
         0          0          0          1
Exporting Registered geometry to kinect/guy/2_aligned.ply...DONE
real 18.66
user 18.54
sys 0.11
$ ../../Super4PCS -i kinect/office/data_02.ply kinect/office/data_03.ply -d 0.03 -o 0.7 -n 1200 -c 100 -r kinect/office/data_03_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1300 points
Initial LCP: 0.115385
Score: 0.530769
(Homogeneous) Transformation from kinect/office/data_03.ply to kinect/office/data_02.ply:
    0.898978     0.279711    -0.337046     0.281543
   -0.297217      0.95481 -0.000359237   -0.0244347
    0.321715     0.100499     0.941488    0.0206971
           0            0            0            1
Exporting Registered geometry to kinect/office/data_03_aligned.ply...DONE
real 13.80
user 13.73
sys 0.06
$ ../../Super4PCS -i kinect/office/data_02.ply kinect/office/data_04.ply -d 0.03 -o 0.7 -n 1200 -c 100 -r kinect/office/data_04_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1124 points
Initial LCP: 0.118327
Score: 0.310498
(Homogeneous) Transformation from kinect/office/data_04.ply to kinect/office/data_02.ply:
 0.829168  0.342566 -0.441735  0.336373
-0.343169  0.935731 0.0815081 0.0823767
 0.441267 0.0840058  0.893435 0.0958307
        0         0         0         1
Exporting Registered geometry to kinect/office/data_04_aligned.ply...DONE
real 6.00
user 5.85
sys 0.14
$ ../../Super4PCS -i kinect/office/data_02.ply kinect/office/data_07.ply -d 0.03 -o 0.7 -n 1200 -c 100 -r kinect/office/data_07_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1333 points
Initial LCP: 0.036009
Score: 0.208552
(Homogeneous) Transformation from kinect/office/data_07.ply to kinect/office/data_02.ply:
   0.592775    0.449729   -0.668103    0.857457
  -0.399008    0.884591    0.241437 -0.00902218
   0.699579    0.123461    0.703808   -0.326819
          0           0           0           1
Exporting Registered geometry to kinect/office/data_07_aligned.ply...DONE
real 37.99
user 37.84
sys 0.12
$ ../../Super4PCS -i kinect/office/data_02.ply kinect/office/data_08.ply -d 0.03 -o 0.7 -n 1200 -c 100 -r kinect/office/data_08_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1287 points
Initial LCP: 0.074592
Score: 0.38073
(Homogeneous) Transformation from kinect/office/data_08.ply to kinect/office/data_02.ply:
  0.863293   0.289121  -0.413684   0.657694
 -0.254516   0.957194   0.137843   0.153694
  0.435829 -0.0137102   0.899925  -0.432885
         0          0          0          1
Exporting Registered geometry to kinect/office/data_08_aligned.ply...DONE
real 29.85
user 29.64
sys 0.18
$ ../../Super4PCS -i kinect/office/data_02.ply kinect/office/data_09.ply -d 0.03 -o 0.7 -n 1200 -c 100 -r kinect/office/data_09_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 1245 points
Initial LCP: 0.029719
Score: 0.44498
(Homogeneous) Transformation from kinect/office/data_09.ply to kinect/office/data_02.ply:
  0.992208   0.109019 -0.0603248   0.729356
 -0.101602   0.988188   0.114726   0.135873
 0.0721195  -0.107703   0.991564  -0.426794
         0          0          0          1
Exporting Registered geometry to kinect/office/data_09_aligned.ply...DONE
real 21.44
user 21.34
sys 0.08
$ ../../Super4PCS -i kinect/stage/1.ply kinect/stage/2.ply -d 0.03 -o 0.5 -n 800 -c 100 -r kinect/stage/2_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 815 points
Initial LCP: 0.087117
Score: 0.51411
(Homogeneous) Transformation from kinect/stage/2.ply to kinect/stage/1.ply:
  0.979288 -0.0937663   0.179455  -0.397404
 0.0741069   0.990794   0.113293   0.119226
 -0.188426 -0.0976479   0.977221  -0.166147
         0          0          0          1
Exporting Registered geometry to kinect/stage/2_aligned.ply...DONE
real 13.77
user 13.42
sys 0.34
$ ../../Super4PCS -i kinect/stage/1.ply kinect/stage/3.ply -d 0.03 -o 0.5 -n 800 -c 100 -r kinect/stage/3_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 815 points
Initial LCP: 0.079755
Score: 0.414724
(Homogeneous) Transformation from kinect/stage/3.ply to kinect/stage/1.ply:
 0.911169  -0.15217  0.382904  -0.91406
0.0690635  0.972561   0.22216 -0.253011
-0.406204 -0.175981  0.896677  0.186795
        0         0         0         1
Exporting Registered geometry to kinect/stage/3_aligned.ply...DONE
real 7.73
user 7.59
sys 0.14
$ ../../Super4PCS -i kinect/stage/1.ply kinect/stage/4.ply -d 0.03 -o 0.5 -n 800 -c 100 -r kinect/stage/4_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 815 points
Initial LCP: 0.074847
Score: 0.31411
(Homogeneous) Transformation from kinect/stage/4.ply to kinect/stage/1.ply:
 0.657959 -0.179931  0.731242  -1.72887
 0.176293  0.980855 0.0827264  0.118251
-0.732128 0.0744825  0.677083  0.312368
        0         0         0         1
Exporting Registered geometry to kinect/stage/4_aligned.ply...DONE
real 16.39
user 16.19
sys 0.19
$ ../../Super4PCS -i kinect/stage/1.ply kinect/stage/5.ply -d 0.03 -o 0.4 -n 800 -c 100 -r kinect/stage/5_aligned.ply  -t 10000
(PLY) ignoring unknown element camera
haveColor
(PLY) ignoring unknown element camera
haveColor
Use Super4PCS
norm_max_dist: 0.030000
Work with 830 points
Initial LCP: 0.077108
Score: 0.3
(Homogeneous) Transformation from kinect/stage/5.ply to kinect/stage/1.ply:
  0.563239  -0.405364   0.720029   -1.83744
  0.186155   0.911245   0.367396  -0.659096
 -0.805052 -0.0728946   0.588709   0.813554
         0          0          0          1
Exporting Registered geometry to kinect/stage/5_aligned.ply...DONE
real 26.90
user 26.54
sys 0.36
```

**Comment by freealong** [comment]
Thanks for your help.

**Comment by nmellado** [comment]
Dear @freealong ,

I may have found the source of your bug.
Could you please try to compile and run the code from branch  `fixUseAsLibrary`, and tell me how it behaves ?
FYI: this branch is related to PR #40, my new release candidate.

Thanks

**Comment by nmellado** [comment]
Dear @freealong,

The runtime stability of the algorithm has been improved with v1.1.1 (better random number management, etc..).
The code now works fine on clang, see CI results here: https://travis-ci.org/nmellado/Super4PCS/jobs/263105579#L1838.

Please let me know if you are still having troubles. The results still might be different for rotated inputs, but however it should be more stable.

Cheers,

**Comment by nmellado** [comment]
I'm closing this issue, it is probably deprecated (published before v1.1.1).
Please re-open a new issue in case the bug is not fixed.
